### PR TITLE
Validate ti rule

### PR DESCRIPTION
**Pull request by manasdk**
* Rule tester now support objects already present in the DB.
* This tool now works with both objects from the file system and objects
  from the DB in any combination.
* Bringing DB support required config-file support, leading to oslo.config,
  leading to the litany of troubles with oslo config and argparse in same app.
  This has evolved into some compromises and a not so friendly help interface
  but the rule-tester is atleast more capable.
* There is some extra logging and rule_tester specific extra_info making the
  output somewhat verbose but it possibly leads to more usable tool.

#### Note
* Using DB implies config now needs to be supplied. This also means lots of ugly log statements unfortunately. Also, even if DB connection is not required the way common_setup work we will enable it anyway. Can try to work around this if is perceived as a blocker.
* Mixing and matching between DB and file system objects is possible but not highlighted.

### From DB on failure
```
./st2reactor/bin/st2-rule-tester --config-file ./conf/st2.dev.conf --rule-ref default.rule4 --trigger-instance-id 566b4be732ed352a09cd3482
2015-12-11 15:24:16,459 INFO [-] Connecting to database "st2" @ "0.0.0.0:27017" as user "None".
2015-12-11 15:24:16,527 INFO [-] Validating rule default.rule4 for st2.generic.actiontrigger.
2015-12-11 15:24:16,542 INFO [-] Validation for rule default.rule4 failed on -
{
    "pattern": "succeeded",
    "type": "iequals",
    "payload_value": "failed",
    "key": "trigger.status"
}
2015-12-11 15:24:16,545 INFO [-] 0 rule(s) found to enforce for st2.generic.actiontrigger.
2015-12-11 15:24:16,546 INFO [-] === RULE DOES NOT MATCH ===
```

### From DB on success
```
$ ./st2reactor/bin/st2-rule-tester --config-file ./conf/st2.dev.conf --rule-ref default.rule4 --trigger-instance-id 566b4be632ed352a09cd347d
2015-12-11 15:24:48,951 INFO [-] Connecting to database "st2" @ "0.0.0.0:27017" as user "None".
2015-12-11 15:24:49,017 INFO [-] Validating rule default.rule4 for st2.generic.actiontrigger.
2015-12-11 15:24:49,030 INFO [-] 1 rule(s) found to enforce for st2.generic.actiontrigger.
2015-12-11 15:24:49,031 INFO [-] === RULE MATCHES ===
```

### From filesystem on failure
```
$ ~/st2/st2reactor/bin/st2-rule-tester --rule=./my_rule.yaml --trigger-instance=./trigger_instance_2.yaml
2015-12-11 15:31:44,959 INFO [-] Connecting to database "st2" @ "0.0.0.0:27017" as user "None".
2015-12-11 15:31:45,025 INFO [-] Validating rule irc.relayed_matched_irc_message for pubmsg.
2015-12-11 15:31:45,041 INFO [-] Validation for rule irc.relayed_matched_irc_message failed on -
{
    "pattern": "StackStorm",
    "type": "icontains",
    "payload_value": "blah blah",
    "key": "trigger.message"
}
2015-12-11 15:31:45,043 INFO [-] 0 rule(s) found to enforce for pubmsg.
2015-12-11 15:31:45,044 INFO [-] === RULE DOES NOT MATCH ===
```

### From filesystem on success
```
$ ~/st2/st2reactor/bin/st2-rule-tester --rule=./my_rule.yaml --trigger-instance=./trigger_instance_1.yaml
2015-12-11 15:32:28,970 INFO [-] Connecting to database "st2" @ "0.0.0.0:27017" as user "None".
2015-12-11 15:32:29,031 INFO [-] Validating rule irc.relayed_matched_irc_message for pubmsg.
2015-12-11 15:32:29,043 INFO [-] 1 rule(s) found to enforce for pubmsg.
2015-12-11 15:32:29,043 INFO [-] === RULE MATCHES ===
```

### CODE REVIEW

**Comment by estee-tew** [comment]
https://stackstorm.slack.com/archives/stackstorm/p1449877485012870

**Comment by manasdk** [comment]
@Kami @lakshmi-kannan do unleash you hate :)

File: docs/source/rules.rst [hunk 1]
```diff
@@ -369,9 +372,49 @@ Output:
...
+    2015-12-11 15:24:16,527 INFO [-] Validating rule my_pack.fire_on_execution for st2.generic.actiontrigger.
+    2015-12-11 15:24:16,542 INFO [-] Validation for rule my_pack.fire_on_execution failed on -
+    {
+        "pattern": "succeeded",
+        "type": "iequals",
+        "payload_value": "failed",
+        "key": "trigger.status"
+    }
+    2015-12-11 15:24:16,545 INFO [-] 0 rule(s) found to enforce for st2.generic.actiontrigger.
+    2015-12-11 15:24:16,546 INFO [-] === RULE DOES NOT MATCH ===
```
Comment (lakshmi-kannan) [inline docs/source/rules.rst]: We'd like to show more details on which criteria didn't match. Are you going to be doing this in another PR?

Comment (manasdk) [inline docs/source/rules.rst]: You mean that log statement is not what solves your concern?

Maybe I can work on the wording.

Comment (manasdk) [inline docs/source/rules.rst]: Few improvement. See if that helps.

**Comment by manasdk** [comment]
Going to merge once everything passes. If you have comments will pick up later.

**Comment by Kami** [comment]
Nice :+1: 

(Looks like I missed this PR, probably when I was traveling :))

File: st2reactor/st2reactor/rules/filter.py [hunk 2]
```diff
@@ -28,7 +28,7 @@
 
 
 class RuleFilter(object):
-    def __init__(self, trigger_instance, trigger, rule):
+    def __init__(self, trigger_instance, trigger, rule, extra_info=False):
```
Comment (Kami) [inline st2reactor/st2reactor/rules/filter.py]: Would be good to add a docstring for `extra_info` argument and maybe also call it `include_extra_info` or smth to make it more clear what is going on :)

File: st2reactor/st2reactor/rules/filter.py [hunk 3]
```diff
@@ -56,7 +57,7 @@ def filter(self):
 
         :rtype: ``bool``
         """
-        LOG.info('Validating rule %s for %s.', self.rule.id, self.trigger['name'],
+        LOG.info('Validating rule %s for %s.', self.rule.ref, self.trigger['name'],
```
Comment (Kami) [inline st2reactor/st2reactor/rules/filter.py]: :+1:
